### ISSUE TITLE
Typescript compilation error: Property 'semver' does not exist on type 'Root'

### ISSUE THREAD

**Issue by momocow** [dszakallas/joi-extension-semver#33]
I met this compilation error in my TypeScript project.

The minimum reproducible snippet is as follows.

- package.json

```json
{
  "name": "test-joi-semver",
  "version": "1.0.0",
  "dependencies": {
    "@hapi/joi": "^17.1.1",
    "joi-extension-semver": "^4.1.1"
  },
  "devDependencies": {
    "@types/hapi__joi": "^17.1.2",
    "typescript": "^3.9.5"
  }
}

```

- index.ts

```ts
import libjoi from '@hapi/joi'
import { semver } from 'joi-extension-semver'

const Joi: libjoi.Root = libjoi.extend(semver)

export default {
  version: Joi.semver()  // <- Property 'semver' does not exist on type 'Root'
}
```

----

I have found out that it's because the inconsistent versions of `@types/hapi__joi`.

In my package.json, `@types/hapi__joi` is required to be v17 because I'm using `@hapi/joi` v17, while it is required to be v16 in `node_modules/joi-extension-semver/package.json` from this package.

The error comes from how NPM deals with the conflicted versions of packages, it create a nested `node_modules` inside the package directory locally, i.e. `node_modules/joi-extension-semver/node_modules/`. Therefore the [declaration merging](https://github.com/dszakallas/joi-extension-semver/blob/c748b28c72b5e2f96bba6ed39652f50cc80429ad/types/index.d.ts#L30-L35) of this package does not take effect since the default type root for TypeScript compiler is `node_modules/@types`.

The declaration from this package is merged into the `Joi.Root` from `node_modules/joi-extension-semver/node_modules/@types/hapi__joi` while the project is compiled against the one from `node_modules/@types/hapi__joi`, so the compilation error is thrown.

---

Possible solution: Define the `@types/hapi__joi` as a peer dependency just like what `@hapi/joi` is.

---

Thank you for the nice package!

**Comment by paulmelnikow** [comment]
One issue with making it a peer dependency is that non-TS users will need to install it, which is a bit irksome for them. Would making it an `optionalDependency` work for you?

**Comment by momocow** [comment]
That's why I think `@types/hapi__joi` should be a peer dependency, **type definitions should come with their sources**.

Since `@hapi/joi` is a peer dependency, it is installed manually by a user with a version matched with `>=16`.

If the actual version of `@hapi/joi` is decided by a user, he should be responsible to it; however, he cannot be responsible to the version of joi's type definitions, `@types/hapi__joi`, currently.

Since the version of `@types/hapi__joi` is not controlled by users, I think everyone, including non-TS users, is possible to meet the version conflict problem (if they manually install it).

A bit of more irksome for them can sometimes save their time on finding why the type hint is not as expected!

---

For a non-TS user, `@types/hapi__joi` is required only for type hints when developing.

In most of cases, this is the responsibility of IDEs, editors, his dev toolchains or himself, and actually VSCode does it for users in the background after they make changes to `package.json`, e.g. `npm install @hapi/joi`.

> Many popular libraries ship with typings files so you get IntelliSense for them automatically. For libraries that do not include typings, VS Code's Automatic Type Acquisition will automatically install community maintained typings file for you.
>
> https://code.visualstudio.com/docs/nodejs/working-with-javascript#_typings-and-automatic-type-acquisition

What's more is that types from local `node_modules` is more preferable by VSCode against the auto-installed types. Users are able to fully control their dependencies and those type definitions.

Let them meet their responsibility if they want to have those type hints.

---

Optional dependency may work, but I have to include `--no-optional` when doing `npm install` in all of my workspaces.

**Comment by paulmelnikow** [comment]
Joi 17 has been renamed back to `joi`, and is now shipping types with the library, so this will be resolved once we upgrade. (#41)

**Comment by paulmelnikow** [comment]
Published as 5.0.0 🎉
